arch/risc-v/src/common/riscv_misaligned.c

```diff
@@ -0,0 +1,448 @@
+        ctx->src  = (uint8_t *)&regs[insn.sdsp.rs2];
+        ctx->len  = 8;
+        break;
+
+      default:
+        _alert("Compressed: %x\n", insn.insn);
+        return false;
+    }
+
+  regs[REG_EPC] += 2;
+
+  return true;
+#else
+  return false;
+#endif
+}
+
+/****************************************************************************
+ * Name: decode_insn_ctx
+ *
+ * Description:
+ *   Try to decode a normal encoding
+ *
+ ****************************************************************************/
+
+static bool decode_insn_ctx(uintptr_t *regs, rv_insn_ctx_t *ctx)
+{
+  uint32_t in;
+  rv_insn_t insn;
+  int32_t imm;
+
+  insn.insn = *(uint32_t *)regs[REG_EPC];
+
+  /* Get load/store instruction encoding */
+
+  in = insn.insn & INSN_MASK;
+
+  switch (in)
+    {
+      case INSN_LH:
+      case INSN_LHU:
+      case INSN_LW:
+      case INSN_LWU:
+      case INSN_LD:
+
+        /* Load a value from memory to register */
+
+        ctx->dest = (uint8_t *)&regs[insn.l.rd];
+        ctx->src = (uint8_t *)regs[insn.l.rs1] + insn.l.imm;
+
+        /* Zero the target register */
+
+        *(uintptr_t *)ctx->dest = 0;
+
+        /* Get data wide bit */
+
+        in &= (~0x4000);
+        in >>= 12;
+        ctx->len = 2 << (in - 1);
+
+        break;
+
+      case INSN_SH:
+      case INSN_SW:
+      case INSN_SD:
+
+        /* Fetch sign extended offset */
+
+        *(uint32_t *)&imm = insn.s.imm2 | (insn.s.imm1 << 5);
+
+        ctx->dest = (uint8_t *)regs[insn.s.rs1] + imm;
+        ctx->src = (uint8_t *)&regs[insn.s.rs2];
+
+        /* Get data wide bit */
+
+        in >>= 12;
+        ctx->len = 2 << (in - 1);
+
+        break;
+
+      case INSN_FLW:
+      case INSN_FLD:
+      case INSN_FSW:
+      case INSN_FSD:
+      default:
+        _alert("Uncompressed: %x\n", insn.insn);
+        return false;
+    }
+
+  /* Adjust EPC */
+
+  regs[REG_EPC] += 4;
+
+  return true;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: riscv_misaligned
+ *
+ * Description:
+ *   This is software interrupt exception handler that handle load/store
+ *   address misaligned exception.
+ *
+ * Input Parameters:
+ *   regs - The current exception context
+ *
+ ****************************************************************************/
+
+int riscv_misaligned(int irq, void *context, void *arg)
+{
+  uintptr_t *regs = context;
+  rv_insn_ctx_t ctx;
+  bool ret;
+
+  /* Try to decode compressed instruction if it is */
+
+  ret = decode_insn_compressed(regs, &ctx);
+
+  /* Decode instruction context */
+
+  if (ret == false)
+    {
+      if (decode_insn_ctx(regs, &ctx) == false)
+        {
+          /* Decode failed, we can't handle a invalid instruction */
+
+          PANIC();
```

Comment:
PANIC() here is fine for now but another option is to return from this function and let the IRQ dispatching logic decide what to do.

In flat mode there is nothing more to do but later with kernel mode we could kill the faulting process. 

But this is fine like it is !

Reply:
Sure, thanks for reviewing